internal/clients/team1/foraging.go

```diff
@@ -0,0 +1,269 @@
+package team1
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+
+	"github.com/sajari/regression"
+)
+
+// ForageOutcome is how much we put in to a foraging trip and how much we got
+// back
+type ForageOutcome struct {
+	turn         uint
+	participant  shared.ClientID
+	contribution shared.Resources
+	revenue      shared.Resources
+}
+
+func (o ForageOutcome) ROI() float64 {
+	if o.contribution == 0 {
+		return 0
+	}
+
+	return float64((o.revenue / o.contribution) - 1)
+}
+
+func (o ForageOutcome) profit() shared.Resources {
+	return o.revenue - o.contribution
+}
+
+type ForageHistory map[shared.ForageType][]ForageOutcome
+
+func (c client) randomForage() shared.ForageDecision {
+	// Up to 10% of our current resources
+	forageContribution := shared.Resources(0.2*rand.Float64()) * c.gameState().ClientInfo.Resources
+	var forageType shared.ForageType
+	if rand.Float64() < 0.5 {
+		forageType = shared.DeerForageType
+	} else {
+		forageType = shared.FishForageType
+	}
+
+	return shared.ForageDecision{
+		Type:         forageType,
+		Contribution: forageContribution,
+	}
+}
+
+func (c client) roiForage() shared.ForageDecision {
+	forageType := bestROIForageType(c.forageHistory)
+	if forageType == shared.ForageType(-1) {
+		// Here we have found that the best idea is to not forage
+		return shared.ForageDecision{
+			Type:         shared.FishForageType,
+			Contribution: 0,
+		}
+	}
+	expectedOutcome := bestROIOutcome(c.forageHistory[forageType])
+
+	contribution := expectedOutcome.contribution
+	// Cap to 20% of resources
+	contribution = shared.Resources(math.Min(
+		float64(contribution),
+		c.config.forageContributionCapPercent*float64(c.gameState().ClientInfo.Resources),
+	))
+	// Add some noise
+	contribution += shared.Resources(math.Min(
+		rand.Float64(),
+		c.config.forageContributionNoisePercent*float64(c.gameState().ClientInfo.Resources),
+	))
+
+	forageDecision := shared.ForageDecision{
+		Type:         forageType,
+		Contribution: contribution,
+	}
+	return forageDecision
+}
+
+func (c client) regressionForage() shared.ForageDecision {
+	bestReward := shared.Resources(0)
+	var decision shared.ForageDecision
+
+	for forageType := range c.forageHistory {
+		r := outcomeRegression(c.forageHistory[forageType])
+		contribution := c.regressionOptimalContribution(r)
+		if contribution > c.gameState().ClientInfo.Resources {
+			contribution = c.gameState().ClientInfo.Resources
+		}
+		expectedRewardF, _ := r.Predict([]float64{float64(contribution)})
+		expectedReward := shared.Resources(expectedRewardF)
+
+		if expectedReward > bestReward {
+			bestReward = expectedReward
+			decision = shared.ForageDecision{
+				Type:         forageType,
+				Contribution: contribution,
+			}
+		}
+	}
+
+	return decision
+}
+
+const flipScale = 0.3
```

Comment:
meaning?

Reply:
It's a scale factor for the contribution of flipForage. It should actually go to the agent config.